parsers/esdtTransferParser.go

```diff
@@ -0,0 +1,181 @@
+		return nil, ErrNotEnoughArguments
+	}
+	esdtTransfers := &vmcommon.ParsedESDTTransfers{
+		ESDTTransfers: make([]*vmcommon.ESDTTransfer, 1),
+		RcvAddr:       rcvAddr,
+		CallArgs:      make([][]byte, 0),
+		CallFunction:  "",
+	}
+	if len(args) > MinArgsForESDTTransfer {
+		esdtTransfers.CallFunction = string(args[MinArgsForESDTTransfer])
+	}
+	if len(args) > MinArgsForESDTTransfer+1 {
+		esdtTransfers.CallArgs = append(esdtTransfers.CallArgs, args[MinArgsForESDTTransfer+1:]...)
+	}
+	esdtTransfers.ESDTTransfers[0] = &vmcommon.ESDTTransfer{
+		ESDTValue:      big.NewInt(0).SetBytes(args[1]),
+		ESDTTokenName:  args[0],
+		ESDTTokenType:  uint32(vmcommon.Fungible),
+		ESDTTokenNonce: 0,
+	}
+
+	return esdtTransfers, nil
+}
+
+func (e *esdtTransferParser) parseSingleESDTNFTTransfer(sndAddr, rcvAddr []byte, args [][]byte) (*vmcommon.ParsedESDTTransfers, error) {
+	if len(args) < MinArgsForESDTNFTTransfer {
+		return nil, ErrNotEnoughArguments
+	}
+	esdtTransfers := &vmcommon.ParsedESDTTransfers{
+		ESDTTransfers: make([]*vmcommon.ESDTTransfer, 1),
+		RcvAddr:       rcvAddr,
+		CallArgs:      make([][]byte, 0),
+		CallFunction:  "",
+	}
+
+	if bytes.Equal(sndAddr, rcvAddr) {
+		esdtTransfers.RcvAddr = args[3]
+	}
+	if len(args) > MinArgsForESDTNFTTransfer {
+		esdtTransfers.CallFunction = string(args[MinArgsForESDTNFTTransfer])
+	}
+	if len(args) > MinArgsForESDTNFTTransfer+1 {
+		esdtTransfers.CallArgs = append(esdtTransfers.CallArgs, args[MinArgsForESDTNFTTransfer+1:]...)
+	}
+	esdtTransfers.ESDTTransfers[0] = &vmcommon.ESDTTransfer{
+		ESDTValue:      big.NewInt(0).SetBytes(args[2]),
+		ESDTTokenName:  args[0],
+		ESDTTokenType:  uint32(vmcommon.NonFungible),
+		ESDTTokenNonce: big.NewInt(0).SetBytes(args[1]).Uint64(),
+	}
+
+	return esdtTransfers, nil
+}
+
+func (e *esdtTransferParser) parseMultiESDTNFTTransfer(sndAddr, rcvAddr []byte, args [][]byte) (*vmcommon.ParsedESDTTransfers, error) {
+	if len(args) < MinArgsForMultiESDTNFTTransfer {
+		return nil, ErrNotEnoughArguments
+	}
+	esdtTransfers := &vmcommon.ParsedESDTTransfers{
+		RcvAddr:      rcvAddr,
+		CallArgs:     make([][]byte, 0),
+		CallFunction: "",
+	}
+
+	numOfTransfer := big.NewInt(0).SetBytes(args[0])
+	startIndex := uint64(1)
+	isTxAtSender := false
+	if bytes.Equal(sndAddr, rcvAddr) {
+		esdtTransfers.RcvAddr = args[0]
+		numOfTransfer.SetBytes(args[1])
+		startIndex = 2
+		isTxAtSender = true
+	}
+
+	minLenArgs := ArgsPerTransfer*numOfTransfer.Uint64() + startIndex
+
+	if uint64(len(args)) > minLenArgs {
+		esdtTransfers.CallFunction = string(args[minLenArgs])
+	}
+	if uint64(len(args)) > minLenArgs+1 {
+		esdtTransfers.CallArgs = append(esdtTransfers.CallArgs, args[minLenArgs+1:]...)
+	}
+
+	esdtTransfers.ESDTTransfers = make([]*vmcommon.ESDTTransfer, numOfTransfer.Uint64())
+	for i := uint64(0); i < numOfTransfer.Uint64(); i++ {
+		tokenStartIndex := startIndex + i*ArgsPerTransfer
+		esdtTransfers.ESDTTransfers[i] = &vmcommon.ESDTTransfer{
```

Comment:
the creation of the new ESDTTransfer instance along with the following if could have been extracted in a new function

Reply:
moved